histogram/src/compact.rs

```diff
@@ -0,0 +1,48 @@
+use serde_derive::{Deserialize, Serialize};
+
+use crate::histogram::Histogram;
+
+/// A `CompactHistogram` is a sparse, columnar representation of the
+/// Histogram. It is significantly smaller than a regular Histogram
+/// when a large number of buckets are zero, which is a frequent
+/// occurence; consequently it is used as the serialization format
+/// of the Histogram. It stores an individual vector for each field
+/// of non-zero buckets. Assuming index[0] = n, (index[0], count[0])
+/// corresponds to the nth bucket.
+#[derive(Default, Serialize, Deserialize)]
+pub struct CompactHistogram {
+    /// parameters representing the resolution and the range of
+    /// the histogram tracking request latencies
+    pub m: u32,
+    pub r: u32,
+    pub n: u32,
+    /// indices for the non-zero buckets in the histogram
+    pub index: Vec<usize>,
+    /// histogram bucket counts corresponding to the indices
+    pub count: Vec<u32>,
+}
+
+impl From<&Histogram> for CompactHistogram {
+    fn from(histogram: &Histogram) -> Self {
+        let mut index = Vec::new();
```

Comment:
Probably doesn't matter, but I wonder if it makes sense to do `Vec::with_capacity(histogram.buckets.len())` to avoid realloc during serialization

Reply:
I don't think that's the right behaviour since we will be filtering out all the zero buckets.

Reply:
I don't think it's harmful or incorrect to reserve enough space to hold all buckets. Since you're pushing into the vec later, the length will still match the number of bucket values you have retained, and the extra capacity (that's uninitialized) won't be serialized out.

Minor thing, we can skip for now.